core/src/main/java/jenkins/mvn/SettingsProvider.java

```diff
@@ -21,12 +27,56 @@
 public abstract class SettingsProvider extends AbstractDescribableImpl<SettingsProvider> implements ExtensionPoint {
 
     /**
+     * Configure maven launcher argument list with adequate settings path.
+     *
+     * <p>Implementations should
+     * <ul>Be aware that this method might get called multiple times during a build.</ul>
+     * <ul>Implement this method. This class provides a default implementation when the given {@code build} is an {@link AbstractBuild}
+     * delegating to {@link #supplySettings(AbstractBuild, TaskListener)} so that implementations have time to adapt.</ul>
+     * </p>
+     *
+     * @param build       the build to provide the settings for
+     * @param workspace the workspace in which the build takes place
+     * @param listener the listener of this given build
+     * @return the filepath to the provided file. {@code null} if no settings will be provided.
+     * @throws IOException typically occur when the {@link #supplySettings(Run, FilePath, TaskListener)} implementation
+     *         accesses to the build environment on the build agent (copying a file to disk...)
+     * @throws InterruptedException typically occur when the {@link #supplySettings(Run, FilePath, TaskListener)} implementation
+     *         accesses to the build environment on the build agent (copying a file to disk...)
+     * @since TODO
+     */
+    @CheckForNull
+    public FilePath supplySettings(@Nonnull Run<?, ?> build,  @Nonnull FilePath workspace, @Nonnull TaskListener listener) throws IOException, InterruptedException {
+        if (build instanceof AbstractBuild && Util.isOverridden(SettingsProvider.class, this.getClass() , "supplySettings",AbstractBuild.class, TaskListener.class)) {
+            AbstractBuild abstractBuild = (AbstractBuild) build;
+            return supplySettings(abstractBuild, listener);
+        } else {
+            throw new AbstractMethodError("Class " + getClass() + " must override the new method supplySettings(Run<?, ?> run, FilePath workspace, TaskListener listener)");
```

Comment:
Is it necessary to throw undocumented Runtime exception here? I would rather write the warning to System logs and return null. And maybe add `assert` for tests (could be useful if you want to create InjectedTest in JTH checking that one of the methods is overridden).

Reply:
Don't you think that returning null will result in strange build results, on the other hand, an exception will fail the build.